lib/netplay/port_mapping_manager.h

```diff
@@ -164,25 +184,31 @@ class PortMappingManager
 	friend void PlumMappingCallbackSuccess(int mappingId, const std::string& externalHost, uint16_t externalPort);
 	friend void PlumMappingCallbackFailure(int mappingId);
 
+	friend void MiniupnpcMappingCallbackSuccess(int mappingId, const std::string& externalHost, uint16_t externalPort);
+	friend void MiniupnpcMappingCallbackFailure(int mappingId);
+
 	// Set discovery status to `SUCCESS`.
 	// The method saves the discovered IP address + port combination inside the
 	// `PortMappingManager` instance so that subsequent `schedule_callback()`
 	// invocations can use this data.
 	//
 	// The method shall only be called from the LibPlum's discovery callback.
-	void resolve_success_fromlibplum(int mappingId, std::string externalIp, uint16_t externalPort);
+	void resolve_success_fromimpl(PortMappingImpl::Type type, int mappingId, std::string externalIp, uint16_t externalPort);
 
 	// Forcefully set discovery status to `status`.
 	// The method shall only be called from LibPlum's discovery callback in case the discovery
 	// process has failed for any reason as reported directly by LibPlum.
 	//
 	// The `status` may only take `FAILURE` or `TIMEOUT` values.
-	void resolve_failure_fromlibplum(int mappingId, PortMappingDiscoveryStatus failureStatus);
+	void resolve_failure_fromimpl(PortMappingImpl::Type type, int mappingId, PortMappingDiscoveryStatus failureStatus);
 
 private:
 
 	~PortMappingManager();
 
+	static std::shared_ptr<PortMappingImpl> getImpl(size_t idx);
+	bool getNextImpl(size_t startingIdx);
```

Comment:
```suggestion
	bool switchToNextWorkingImpl(PortMappingImpl::Type startingTypeIdx);
```
This method name should somehow tell us that it switches the active implementation to the next one, so I think a name change is needed.